src/push.go

```diff
@@ -53,7 +58,54 @@ func (f *PushOnlyFlags) Validate() Validations {
 	return validations
 }
 
+func GetToken(ctx context.Context, flags *PushFlags) (*string, error) {
+	fmt.Printf("Getting an impersonation token for `%s`..\n", flags.ActionsAdminUser)
+
+	ts := oauth2.StaticTokenSource(&oauth2.Token{AccessToken: flags.Token})
+	tc := oauth2.NewClient(ctx, ts)
+	ghClient, err := github.NewEnterpriseClient(flags.BaseURL, flags.BaseURL, tc)
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating enterprise client")
+	}
+
+	rootRequest, err := ghClient.NewRequest("GET", enterpriseAPIPath, nil)
+	if err != nil {
+		return nil, errors.Wrap(err, "Error constructing request for GitHub Enterprise client.")
+	}
+	rootResponse, err := ghClient.Do(ctx, rootRequest, nil)
+	if err != nil {
+		return nil, errors.Wrap(err, "Error checking connectivity for GitHub Enterprise client.")
+	}
+
+	isAE := rootResponse.Header.Get(enterpriseVersionHeaderKey) == enterpriseAegisVersionHeaderValue
+
+	minimumRepositoryScope := "public_repo"
+	if isAE {
+		// the default repository scope for non-ae instances is 'public_repo'
+		// while it is `repo` for ae.
+		minimumRepositoryScope = "repo"
+		fmt.Printf("Running against GitHub AE, changing the repository scope to '%s'..\n", minimumRepositoryScope)
+	}
+
+	impersonationToken, _, err := ghClient.Admin.CreateUserImpersonation(ctx, flags.ActionsAdminUser, &github.ImpersonateUserOptions{Scopes: []string{minimumRepositoryScope, "workflow"}})
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to impersonate Actions admin user.")
+	}
+
+	fmt.Printf("Got the impersonation token for `%s`..\n", flags.ActionsAdminUser)
+	var result = impersonationToken.GetToken()
+	return &result, nil
+}
+
 func Push(ctx context.Context, flags *PushFlags) error {
+	var token, err = GetToken(ctx, flags)
```

Comment:
Should there be an option for not using an impersonation token if you're not trying to sync a repository in the `actions` organization? Maybe something like `--actions-admin-user -` to indicate you just want to use the token provided directly.

Reply:
A very good point. I am looking into how can I try to automagically avoid the impersonation unless it's really required. But it's not as straightforward. 

And I will reach out in DMs about the flag syntax you proposed.

Reply:
This is a good point, where customers doesnt have to impersonate as actions-admin when they are trying to sync the repos to different orgs.

With the flag `--actions-admin-user -`, this makes those customers to provide this option always incase they are trying for different orgs. More or less like a required field. I feel like this breaks the current customer experience.

What do you think ?

we can go with 2 options:
- if customer doesnt provide `--actions-admin-user`, don't impersonate. If they want, let them provide the actions-admin name.
- OR Introduce a boolean flag like `impersonate-actions-admin (default: false)` and `actions-admin-user is defaulted to actions-admin` unless they override.

Reply:
Sure, I don't mind changing this. I can add a description to the switch and note that the default user should be `actions-admin` and if a user wants to use the default impersonation, they should just type that in explicitly.

Reply:
@ajaykn please see if you like the new version better

Reply:
Looks good thanks you.